File: task-merge/src/main/scala/com/zxkw/userprofile/app/TaskMerge.scala
package com.zxkw.userprofile.app

import java.util.Properties
import com.zxkw.userprofile.common.bean.TagInfo
import com.zxkw.userprofile.common.dao.TagInfoDAO
import com.zxkw.userprofile.common.util.MyPropertiesUtil
import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession

object TaskMerge {

//  1  创建标签宽表
//    一次创建还是 周期性创建
//    标签宽表的列取决于有多少个标签 ，所以每天的列 是动态变化
//    每天都要新建一个标签表    标签表（周期性表）： 表名前缀_日期
//  2  通过多个标签表
//    拼接为标签高表  ， 再利用pivot 进行行转列 转为标签宽表
//  3  把数据写入到宽表中

  def main(args: Array[String]): Unit = {
    //  1  创建标签宽表
    //    一次创建还是 周期性创建
    //    标签宽表的列取决于有多少个标签 ，所以每天的列 是动态变化
    //    每天都要新建一个标签表    标签表（周期性表）： 表名前缀_日期
    // create table tableName
    //  ( uid string , tag_code1  string , tag_code2 string ....)
    //  partitioned   因为每天建表 所以 不同分区
    //    ROW FORMAT DELIMITED FIELDS TERMINATED BY '\\t'
    //   LOCATION  '$hdfsStorePath/$userProfileDbName/$tableName'
    val sparkConf: SparkConf = new SparkConf().setAppName("task_merge_app")//.setMaster("local[*]")
    val sparkSession: SparkSession = SparkSession.builder().config(sparkConf).enableHiveSupport().getOrCreate()


    val taskId: String = args(0)
    val taskDate: String = args(1)

    val tableName=s"user_tag_merge_${taskDate.replace("-","")}"
    // 标签宽表的列 取决于 有多少个 启用状态的标签 --> 通过查询mysql中的tag_info+task_info
    val tagInfoList: List[TagInfo] = TagInfoDAO.getTagInfoListTaskOn()
    val tagCodeList: List[String] = tagInfoList.map(tagInfo=> s" ${tagInfo.tagCode.toLowerCase} string")
    val tagCodeSQL: String = tagCodeList.mkString(",")


    val properties: Properties = MyPropertiesUtil.load("config.properties")
    val userProfileDbName = properties.getProperty("user-profile.dbname")
    val wareHouseDbName = properties.getProperty("data-warehouse.dbname")
    val hdfsStorePath = properties.getProperty("hdfs-store.path")

   // 考虑幂等性  数据重试计算
    sparkSession.sql(s"drop table if exists  $userProfileDbName.$tableName")

    val createTableSQL=
      s"""
         |
         |      create  table if not exists $userProfileDbName.$tableName
         |      ( uid string , $tagCodeSQL )
         |      ROW FORMAT DELIMITED FIELDS TERMINATED BY '\\t'
         |     LOCATION  '$hdfsStorePath/$userProfileDbName/$tableName'
       """.stripMargin
    println(createTableSQL)
    sparkSession.sql(createTableSQL)


    //  2  通过多个标签表
    //    拼接为标签高表 union all  ， 再利用pivot 进行行转列 转为标签宽表
    // select * from
    // ( select uid ,'tag_code1' tag_code ,tag_value  from  tag_code1 where dt='taskDate'
    //  union all
    //  select uid ,'tag_code2' tag_code ,tag_value  from  tag_code2 where dt='taskDate'
    //  union all
    //  ...) tagtable
    //  pivot (     max( tag_value  ) as  tv   for     tag_code in ('tagcode1' ,'tagcode2','tagcode3'.. )     )


    //select uid ,'tag_code1' tag_code ,tag_value  from  tag_code1 where dt='taskDate'
    //     union all
    //      select uid ,'tag_code2' tag_code ,tag_value  from  tag_code2 where dt='taskDate'
    //     union all
    //     ...
    val tagSQLlist:List[String]=  tagInfoList.map { tagInfo =>
      s"""
         |select uid ,'${tagInfo.tagCode}' tag_code ,tag_value
         | from $userProfileDbName.${tagInfo.tagCode.toLowerCase}
         | where dt='$taskDate'
       """.
        stripMargin
      }
    val tagTableSQL: String = tagSQLlist.mkString(" union all ")

    //'tagcode1' ,'tagcode2','tagcode3'..
    val tagCodePivotSQL: String = tagInfoList.map(tagInfo=>s"'${tagInfo.tagCode}'").mkString(",")


     val  selectSQL=
       s"""
          |select * from
          |($tagTableSQL)
          |    tagtable
          |  pivot (     max( tag_value  ) as  tv   for  tag_code in ( $tagCodePivotSQL ))
        """.stripMargin
     println(selectSQL)


    //  3  把数据写入到宽表中

    val insertSQL=s" insert overwrite table $userProfileDbName.$tableName $selectSQL"
    println(insertSQL)
    sparkSession.sql(insertSQL)


  }





}

File: task-merge/src/main/java/com/zxkw/userprofile/app/TaskMerge.java
package com.zxkw.userprofile.app;

import com.zxkw.userprofile.common.bean.TagInfo;
import com.zxkw.userprofile.common.constants.ConstCode;
import com.zxkw.userprofile.common.dao.TagInfoDao;
import com.zxkw.userprofile.common.util.MyPropsUtil;
import org.apache.commons.lang.StringUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.sql.SparkSession;

import java.util.List;
import java.util.stream.Collectors;

/**
 * 合并标签
 * 1. 取得计算的业务日期
 * 2. 查询Mysql得出所有的需要被合并的标签，其中使用tag_code作为列名
 * 3. 建立宽表, 不能手工建立， 因为不确定有多少列， 每天可能不一样
 * 4. 通过一条pivot SQL语句把多个标签单表合并成为一个标签宽表
 *        维度列： uid  旋转列: tag_code    聚合列: tag_value
 */
public class TaskMerge {
    public static void main(String[] args) {
        SparkConf sparkConf = new SparkConf().setAppName("task_merge_app") .setMaster("local[*]");
        SparkSession sparkSession = SparkSession.builder().config(sparkConf).enableHiveSupport().getOrCreate();


        //1.取得计算的业务日期
        String taskId = args[0];
        String businessDate = args[1];

        //2. 查询Mysql得出所有的需要被合并的标签，其中使用tag_code作为列名
        List<TagInfo> tagInfoList = TagInfoDao.getTagInfoList();

        //3. 建立宽表, 不能手工建立， 因为不确定有多少列， 每天可能不一样
        //  表名?  字段?  分区?  存储位置?  格式?
        // create table if not exists up_tag_merge_[businessDate]
        // (
        // uid string ,
        // [tagCodeSQL(tagCode string )]
        // )
        // row format delimited fields terminated by  '\\t'
        // location '[HDFS_PATH]/[UP_DBNAME]/[TABLENAME]'

        String tableName = "up_tag_merge_" + businessDate.replace("-" , "");
        String hdfsPath = MyPropsUtil.get(ConstCode.HDFS_STORE_PATH);
        String upDbName = MyPropsUtil.get(ConstCode.USER_PROFILE_DBNAME);

        List<String> tagCodeSQLList =
                tagInfoList.stream().map(tagInfo -> tagInfo.getTagCode().toLowerCase() + " string ").collect(Collectors.toList());
        String tagCodeSQL = StringUtils.join(tagCodeSQLList, ",");
        String dropTableSQL = " drop table if exists " + upDbName + "." + tableName;
        System.out.println("dropTableSQL : " + dropTableSQL );
        sparkSession.sql(dropTableSQL);
        String  createTableSQL =  " create table if not exists " + upDbName+ "." + tableName +
                " (" +
                " uid string, "  + tagCodeSQL +
                " )" +
                " row format delimited fields terminated by '\\t'" +
                " location '"+hdfsPath+ "/" + upDbName +"/" + tableName + "'";

        System.out.println("createTableSQL : " + createTableSQL);
        sparkSession.sql(createTableSQL);

        //4. 通过一条pivot SQL语句把多个标签单表合并成为一个标签宽表
        // 维度列： uid  旋转列: tag_code    聚合列: tag_value
        List<String> tagSqlList = tagInfoList.stream().map(
                tagInfo -> "select uid , '"
                        + tagInfo.getTagCode().toLowerCase() +
                        "' as tag_code , tag_value from " +
                        upDbName + "." + tagInfo.getTagCode().toLowerCase() +
                        " where dt = '" + businessDate + "'").collect(Collectors.toList());
        String unionSQL = StringUtils.join(tagSqlList, " union all ");

        List<String> tagCodeList =
                tagInfoList.stream().map(tagInfo -> "'" + tagInfo.getTagCode().toLowerCase() + "'").collect(Collectors.toList());
        String tagCodeQuerySQL = StringUtils.join(tagCodeList, ",");

        String pivotSQL = " select * from ( " + unionSQL + ") "+
                " pivot (max(tag_value) tv for tag_code in ("+ tagCodeQuerySQL+"))" ;

        System.out.println("pivotSQL: " + pivotSQL);

        String insertSQL = " insert overwrite table "+ upDbName+ "."+ tableName + " " + pivotSQL ;
        System.out.println("insertSQL :" + insertSQL);

        sparkSession.sql(insertSQL);

    }
}
